### PR TITLE
update: allow updating from local branch

### PR DESCRIPTION
**Pull request by vladshablinsky**
Allows updating from local branch. For example you can change some formulae on your local `branch_name` and test things like formula-renames, report format and other stuff that is performed on update when some formulae get changed.

The option for this `update from branch` behaviour is `--from-branch`, so if you want to check whether migration occurs on brew update for renamed formula on your branch, you can use `brew update --from-branch`.

Solves https://github.com/Homebrew/homebrew/issues/43472.

### CODE REVIEW

File: Library/Homebrew/cmd/update.rb [hunk 1]
```diff
@@ -216,6 +216,20 @@ def pull!(options = {})
       @initial_branch = ""
     end
 
+    # used for testing purpose, for example for testing formula migrations when
+    # some formula gets renamed on your branch and you test migrations by
+    # running 'brew update --from-branch"
+    if ARGV.include?("--from-branch")
+      @initial_revision = `git rev-parse -q --verify master`.chomp
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/update.rb]: Maybe replace the hard-coded `master` with `#{@upstream_branch}` so it won't fail for the rare cases without a master branch (see #43865)? Same for the `odie` message below.

File: Library/Homebrew/cmd/update.rb [hunk 2]
```diff
@@ -216,6 +216,20 @@ def pull!(options = {})
       @initial_branch = ""
     end
 
+    # used for testing purpose, for example for testing formula migrations when
+    # some formula gets renamed on your branch and you test migrations by
+    # running 'brew update --from-branch"
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/update.rb]: I'd rephrase the comment as follows to make it a bit clearer (I'm not a native speaker): “Used for testing purposes, e.g., for testing formula migrations in the currently checked-out branch. To test, run `brew update --from-branch`.”

**Comment by UniqMartin** [comment]
I'm a bit conflicted about the option name `--from-branch`, because in my mind it's an update *from* master *to* the local branch. What do you think about `--to-current-branch` (my favorite), `--to-local-branch`, or `--to-HEAD` instead?

**Comment by vladshablinsky** [comment]
>I'm a bit conflicted about the option name --from-branch, because in my mind it's an update from master to the local branch. What do you think about --to-current-branch (my favorite), --to-local-branch, or --to-HEAD instead?

I think I disagree here. When we get updates from the upstream we don't say "to upstream". I think something similar we can see here i.e. we get the diff from the current head and use it in report.

In the independent case when we update just from master to origin/master "to" have more sense I think. But in "--from-branch" I mean get the chages `from` it, but not update from some commit `to` it. @UniqMartin what do you think. Doesn't it have sense?

**Comment by UniqMartin** [comment]
> What do you think? Doesn't it make sense?

Thanks! Now I understand why you named it like that and it makes sense. It feels right to say “get the updates *from* a certain branch” or “get the updates *from* upstream”. I'm happy with any of the suggested option names, including `--from-branch`.

**Comment by mikemcquaid** [comment]
I wonder about the argument naming too as it wasn't obvious to me what it was doing. Maybe `--simulate-from-current-branch` or something as it's not actually doing the update?

**Comment by vladshablinsky** [comment]
@mikemcquaid are you ok with `--simulate-from-current-branch`? It describes the purpose more clearly. It's pretty long though.

**Comment by mikemcquaid** [comment]
@vladshablinsky Works for me :+1:
